### ISSUE TITLE
[HOLD on #16660][$1000] Web – Assign Task – Whole page jitter on check and uncheck of checkbox.

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#22656]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open any " Assign Task " chat.
2. Check and uncheck checkbox under title multiple times.
3. Notice after few times of check and uncheck whole page starts to jitter.

## Expected Result:
Page should not start to jitter upon check and uncheck of checkbox.

## Actual Result:
 Page starts to jitter upon selection check and uncheck of checkbox.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.39-5
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/8401f72f-a157-44fc-84f3-75f2a2c30500

https://github.com/Expensify/App/assets/93399543/f4d8680e-7b8e-487a-9d96-2af02d4a2a36

**Expensify/Expensify Issue URL:**
**Issue reported by:** @usmantariq96
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1689069224351539

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0156490fa397b075d8</li>
        <li>Upwork Job ID: 1681396985675022336</li>
        2023-07-25<li>Automatic offers: </li>
<ul>
<li> |  | 0</li>
<li> |  | 0</li>
</ul></li>
    </ul>
</details>

**Comment by michaelhaxhiu** [comment]
This was tough to reproduce, but I got it eventually. Make sure to check/uncheck the task until there is a scroll bar in the thead, then scroll up so that you can't see new updates at the footer if you continue to check/uncheck

<img width="824" alt="image" src="https://github.com/Expensify/App/assets/13206896/23f9a3fc-ddc0-4341-b132-5613723f0c42">

**Comment by robertKozik** [comment]
Indeed very tough to repro, I'll keep it under on radar and wait for any proposals 👌🏼

**Comment by akamefi202** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
In task page, the page jitters when the user clicks(check & uncheck) the checkbox continuously. (Approximately, more than 10~14 times)

### What is the root cause of that problem?
https://github.com/Expensify/App/blob/8dcad168d11a71424b6ffd7fa30fd1770c0165d5/src/components/InvertedFlatList/BaseInvertedFlatList.js#L136-L138
I think it's because of `getItemLayout` in `FlatList`.
Currently, we save size & offset of each item using `measureItemLayout` function while rendering and we input them using `getItemLayout` function in next rendering.
We use `sizeMap` array to save size & offset of items.

But all these code was assumed and written as if new item will be added to the last element of the item array in `FlatList` when the user adds a comment.
Actually, it will be added to first element of the item array.

### What changes do you think we should make in order to solve the problem?
We need to update `getItemLayout` and `measureItemLayout` functions based on correct mechanism.
`const sizeMapIndex = this.props.data.length - index - 1;`
We need to get index again in reverse order when we read & write size information of items with `sizeMap` array.
And also offset of newly added item is 0.

### What alternative solutions did you explore? (Optional)
We may not use `getItemLayout` in `FlatList`. But it could affect performance for list of many items.

**Comment by robertKozik** [comment]
Hi @akamefi202, thanks for your proposal. First of all, I've tried to test your approach on my end, but it causes an error for me. I've attached a diff which I used.

Later, I'm not sure about the root cause you have written. As, most likely, you pinpointed the faulty code correctly, but you didn't provide an answer as to why it only occurs after spamming check/uncheck. If we were wrongly saving measurements, shouldn't it cause problems way earlier, not only after clicking multiple times?

```diff
diff --git a/src/components/InvertedFlatList/BaseInvertedFlatList.js b/src/components/InvertedFlatList/BaseInvertedFlatList.js
index d3fcda0ea5..e941933cec 100644
--- a/src/components/InvertedFlatList/BaseInvertedFlatList.js
+++ b/src/components/InvertedFlatList/BaseInvertedFlatList.js
@@ -53,13 +53,14 @@ class BaseInvertedFlatList extends Component {
      * @return {Object}
      */
     getItemLayout(data, index) {
-        const size = this.sizeMap[index];
+        const sizeMapIndex = this.props.data.length - index - 1;
+        const size = this.sizeMap[sizeMapIndex];
 
         if (size) {
             return {
                 length: size.length,
                 offset: size.offset,
-                index,
+                index: sizeMapIndex,
             };
         }
 
@@ -76,7 +77,7 @@ class BaseInvertedFlatList extends Component {
             // initialRowHeight since we can only assume that all previous items
             // have not yet been measured
             offset: _.isUndefined(lastMeasuredItem) ? this.props.initialRowHeight * index : lastMeasuredItem.offset + this.props.initialRowHeight,
-            index,
+            index: sizeMapIndex,
         };
     }
 
@@ -88,20 +89,22 @@ class BaseInvertedFlatList extends Component {
      */
     measureItemLayout(nativeEvent, index) {
         const computedHeight = nativeEvent.layout.height;
+        const sizeMapIndex = this.props.data.length - index - 1;
 
         // We've already measured this item so we don't need to
         // measure it again.
-        if (this.sizeMap[index]) {
+        if (this.sizeMap[sizeMapIndex]) {
             return;
         }
 
-        const previousItem = this.sizeMap[index - 1] || {};
+        const previousItem = this.sizeMap[sizeMapIndex - 1] || {};
 
         // If there is no previousItem this can mean we haven't yet measured
         // the previous item or that we are at index 0 and there is no previousItem
         const previousLength = previousItem.length || 0;
         const previousOffset = previousItem.offset || 0;
-        this.sizeMap[index] = {
+        this.sizeMap[sizeMapIndex] = {
             length: computedHeight,
             offset: previousLength + previousOffset,
         };
```

**Comment by akamefi202** [comment]
@robertKozik 
About the code changes, we should use original `index` when return size data. We should use `sizeMapIndex` only for reading & writing `sizeMap` array.
```
if (size) {
     return {
         length: size.length,
         offset: size.offset,
         index,
     };
}
```
And `offset` should be 0, when we return size data of newly added item.

About root cause, `getItemLayout` is an optional optimization to improve performance of `FlatList` when it contains so many items.
When `FlatList` has only few items, it is enough to compute size of items correctly.
After many items are added to the list, the component start to refer the size data(which is not accurate now) to reduce time for the measurement of dynamic content.
So we see this issue after many items are added to the list.

One more thing is we're adding new items to invisible viewport when we check & uncheck the checkbox.
I think this is the reason why we see this issue only in task page.
The component gives priority to compute size of visible items and we're adding new items to invisible viewport.
So it might use wrong size data from `sizeMap` array instead of computing size of new items.

This is just my analysis and other engineer might find more accurate reason.
But it's definitely clear that this issue is occurred by `getItemLayout` and we can fix by updating `getItemLayout` and `measureItemLayout` functions.

**Comment by kosmydel** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
After checking/unchecking the task the page jitter.

### What is the root cause of that problem?
The heights/offsets of the elements in the inverted `FlatList` are not calculated correctly. The full explanation for that is [here](https://github.com/necolas/react-native-web/pull/2412). Previously, we addressed it by creating a `BaseInvertedFlatList` component, but it doesn't work fully as expected.

### What changes do you think we should make in order to solve the problem?
Previously, we used the `BaseInvertedFlatList` component for inverted lists, to make it work on the web. However, the `react-native-web` library has addressed this [issue](https://github.com/necolas/react-native-web/issues/1254) related to the `inverted` FlatLists in [this PR](https://github.com/necolas/react-native-web/pull/2501).


**Solution:**
1. We can update the `@expensify/react-native-web` with the latest changes from the `necolas/react-native-web` [repository](https://github.com/necolas/react-native-web).
2. Remove the `BaseInvertedFlatList` component.
3. Replace `<BaseInvertedFlatList />` usages with:
```javascript
<FlatList
 inverted
 {...otherProps}
/>
```

However, there are some dependency-related issues here.

### What alternative solutions did you explore? (Optional)
We can add to the `@expensify/react-native-web` fork changes only from [this PR](https://github.com/necolas/react-native-web/pull/2501/files) and replace `BaseInvertedFlatList` with `FlatList` with `inverted` option.

I created a short _PoC_ video showing the results.

<details>
<summary>
PoC video
</summary>

https://github.com/Expensify/App/assets/104823336/ec4b725e-6d07-440f-8a66-94008572c8e6

</details>

**Comment by robertKozik** [comment]
@akamefi202 Your proposal has some strange behaviour when the message list is very long:

https://github.com/Expensify/App/assets/61146594/a32dabd7-e281-4e91-a9f8-c72bd4b278d2

@kosmydel I think we have bumping the `react-native-web` version in plans. If thats truly the case your proposal is better suited for this issue

**Comment by robertKozik** [comment]
Working on proposals 👀

**Comment by shubham1206agra** [comment]
@robertKozik Can we try to move to flashlist / recyclerlistview instead of using flatlist?

**Comment by stitesExpensify** [comment]
I think that @kosmydel has the best proposal here, and we should wait for the react-native-web update before proceeding. That is in [this issue](https://github.com/Expensify/App/issues/16660) which is also blocked, but we are very close to it being unblocked, so I think it's worth the wait.

**Comment by michaelhaxhiu** [comment]
Since this is on Hold, I'm gonna apply `Weekly` label.

Note: I'm going to be OOO the next ~2 weeks and given this is on Hold, I'm not going to assign a BZ teammate because I feel like this won't be complete before I return. If I'm mistaken and this sees progress before I return, can you please remove my assignment & re-apply the `Bug` label to assign a BZ that's online? Thanks!

**Comment by stitesExpensify** [comment]
Still on hold

**Comment by stitesExpensify** [comment]
hold

**Comment by adelekennedy** [comment]
linked issue is still in progress

**Comment by kosmydel** [comment]
Hey, as the `react-native-web` was merged I had a look at this issue.

I believe, that my original proposal is out of date, as the app has changed significantly. New bugs appeared, which didn't occur when I originally wrote the proposal. 

The report item list is completely broken (on both production and staging):
<details>
<summary>Production</summary>

https://github.com/Expensify/App/assets/104823336/50290451-37fa-4c27-b026-bc75d2d6dd3f


</details>

<details>
<summary>Staging</summary>

https://github.com/Expensify/App/assets/104823336/e92c62be-2ef0-4ec0-848e-f146cf0f52ee


</details>

The flat list looks like is being rewritten from scratch:
- https://github.com/Expensify/App/pull/28793 (it got reverted)
- https://github.com/Expensify/App/issues/23223

Therefore, I think that this issue might not be relevant anymore, as this part of the app is being completely changed.

cc: @allroundexperts

**Comment by kosmydel** [comment]
Hey,
I've investigated this issue and it looks like it is fixed now. [This PR](https://github.com/Expensify/App/pull/31288) introduced changes that I suggested in [my proposal](https://github.com/Expensify/App/issues/22656#issuecomment-1649528599). I've retested this issue and for me, it doesn't occur anymore. I think we can close this one out.

<details>
<summary>
Video
</summary>

https://github.com/Expensify/App/assets/104823336/99bf071e-6282-4e0f-a794-22bb83ed5edf

</details>

cc @allroundexperts, @stitesExpensify, @adelekennedy

**Comment by kosmydel** [comment]
So in that case, can we close this issue?
cc @adelekennedy, @allroundexperts, @stitesExpensify
